File: 2024/December/NumberTheory/BJ2023.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <string>
using namespace std;

// 소수인지 판별하는 함수
bool isPrime(int num) {
    // 1은 소수가 아님
    if (num <= 1) return false;

    // 2와 3은 소수임
    if (num == 2 || num == 3) return true;

    // 짝수는 소수가 아님
    if (num % 2 == 0) return false;

    // 소수 여부 판단을 위해 √num까지만 나눠서 확인
    for (int i = 3; i <= sqrt(num); i += 2) {
        if (num % i == 0) {
            return false;
        }
    }

    return true;
}

int main() {
    int num;
    cin >> num;

    // 메모이제이션을 위한 벡터 배열
    vector<vector<int>> memoObject(num + 1);

    // 1자리 수 소수 초기화
    memoObject[1] = {2, 3, 5, 7};

    // 2자리 수부터 num자리 수까지 소수를 구하는 로직
    for (int i = 2; i <= num; i++) {
        for (const int& beforeNum : memoObject[i - 1]) {
            for (int j = 0; j <= 9; j++) {
                int newNumber = stoi(to_string(beforeNum) + to_string(j));
                if (isPrime(newNumber)) {
                    memoObject[i].push_back(newNumber);
                }
            }
        }
    }

    // 결과 출력
    for (const int& printNumber : memoObject[num]) {
        cout << printNumber << endl;
    }

    return 0;
}

File: 2024/December/NumberTheory/BJ2023.js
const fs = require('fs');
const filepath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const num = Number(fs.readFileSync(filepath).toString().trim());
const memoObject = {};
for (let i = 1; i <= num; i++) {
  memoObject[i] = [];
}

memoObject[1] = [2, 3, 5, 7];

// 소수인지를 판별하는 함수를 작성
// 기존의 숫자가 소수이면, 하나를 더 붙였을 경우만 판단하면 계산 리소스를 줄일 수 있음
// 소수를 판단하는 법
// 1. 2와 3 거르기 2. 짝수 거르기 3. 3부터 자기 자신의 제곱근까지 돌려서 나누어 떨어지는지를 확인.

function isPrime(num) {
  // 1은 소수가 아니므로 false 반환
  if (num <= 1) return false;

  // 2와 3은 소수이므로 true 반환
  if (num === 2 || num === 3) return true;

  // 짝수는 소수가 아니므로 false 반환
  if (num % 2 === 0) return false;

  // num이 2 또는 3으로 나누어지지 않는 경우
  // 소수 여부 판단을 위해 √num까지만 나눠서 확인
  for (let i = 3; i <= Math.sqrt(num); i += 2) {
    if (num % i === 0) {
      return false;
    }
  }

  return true;
}

for (let i = 2; i <= num; i++) {
  for (const beforeNum of memoObject[i - 1]) {
    for (let j = 0; j <= 9; j++) {
      const newNumber = Number(String(beforeNum).concat(String(j)));
      if (isPrime(newNumber)) {
        memoObject[i].push(newNumber);
      }
    }
  }
}

for (const printNumber of memoObject[num]) {
  console.log(printNumber);
}
